core/task_queue.rs

```diff
@@ -0,0 +1,142 @@
+use std::collections::LinkedList;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+
+use crate::futures::task::AtomicWaker;
+use crate::futures::Future;
+use crate::parking_lot::Mutex;
+
+#[derive(Default)]
+struct TaskQueueTaskWaker {
+  is_ready: AtomicBool,
+  waker: AtomicWaker,
+}
+
+#[derive(Default)]
+struct TaskQueueTasks {
+  is_running: bool,
+  wakers: LinkedList<Arc<TaskQueueTaskWaker>>,
+}
+
+/// A queue that executes tasks sequentially one after the other
+/// ensuring order and that no task runs at the same time as another.
+#[derive(Clone, Default)]
+pub struct TaskQueue {
```

Comment:
I have no idea why, but `Semaphore::new(1)` was not as reliable as this at ensuring order. See https://github.com/denoland/deno/pull/15092#discussion_r914857542